include/dr/mhp/algorithms/algorithms.hpp

```diff
@@ -116,16 +116,16 @@ void for_each(DI first, DI last, auto op) {
 //
 
 /// Collective iota on iterator/sentinel for a distributed range
-template <lib::distributed_iterator DI>
-void iota(DI first, DI last, auto value) {
+template <lib::distributed_iterator DI, std::integral T>
```

Comment:
integral is too strong. Won't iota work for float? It is implementable for any type, but some types will not permit parallel execution.

Reply:
The ranges version of `iota` actually doesn't work for float, since `weakly_incrementable` requires that the difference type be an integer-like type.  (For float it's a float, which is not integer-like.)

We could implement a more generic version, but it would be a lot of work.  Will people really want to use `iota` with things that are not integers?

`std::integral` is more restrictive than `std::weakly_incrementable`---the main thing I need from `integral` is the ability to add, rather than just increment, which is necessary for parallelism.  (`x + 12` rather than just `++x`)

Reply:
I made it consistent with shp. @rscohn2, please decide if I should revert `std::integral` in mhp. If revert, then we need to have tests with floats too in mhp.

Reply:
`std::integral` make sense. It should be possible to implement iota on top of `xhp::copy` or `xhp::for_each`. They only support random access iterators, which is necessary for parallelism, so iota could only support integral.

Long term, I think we need to support non random access iterators for `xhp::for_each`. It won't be parallel, but`std::for_each` would be doing element-wise remote memory operations, which is probably 1000x slower.

Reply:
There will potentially be iterators that are not exactly random access but support parallelism, so I agree that we should think about supporting non-random access iterators in the future.  (writing code generically and using `rng::distance`, etc.)

For example, I could imagine implementing a hash table---while the hash table iterator would not exactly be random access, we could iterate through the hash table in parallel on the GPU/across multiple processes.  (e.g. by splitting up the range recursively or implementing a custom `for_each` for the iterator type.)

In the meantime, though, I think it's reasonable to restrict to random access, since all of the data structures we're currently considering have random access (-ish) iterators, including the sparse matrix.